File: ml-service/routes/categorizer.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import logging

router = APIRouter()
logger = logging.getLogger(__name__)

# Category keywords mapping
CATEGORY_KEYWORDS = {
    "Food": ["restaurant", "cafe", "coffee", "pizza", "burger", "food", "lunch", "dinner", "breakfast", "grocery", "supermarket", "market"],
    "Transportation": ["bus", "taxi", "uber", "gas", "petrol", "car", "parking", "train", "flight", "airline"],
    "Entertainment": ["movie", "cinema", "concert", "game", "gaming", "spotify", "netflix", "entertainment", "music"],
    "Shopping": ["mall", "store", "shop", "amazon", "ebay", "fashion", "clothes", "shoes", "retail"],
    "Utilities": ["electricity", "water", "gas", "internet", "phone", "bill", "utility"],
    "Education": ["school", "college", "university", "course", "book", "tuition", "learning"],
    "Health": ["hospital", "doctor", "pharmacy", "medicine", "health", "gym", "fitness"],
    "Rent": ["rent", "apartment", "lease", "housing"],
    "Entertainment & Games": ["game", "gaming", "playstation", "xbox"],
    "Other": []
}

class CategorizationRequest(BaseModel):
    description: str
    confidence_threshold: float = 0.5

class CategorizationResponse(BaseModel):
    category: str
    confidence: float
    keywords_matched: list

def categorize_expense(description: str) -> tuple:
    """
    Categorize expense based on description using keyword matching
    Returns: (category, confidence, matched_keywords)
    """
    description_lower = description.lower()
    scores = {}
    matched_keywords = {}

    for category, keywords in CATEGORY_KEYWORDS.items():
        matches = [k for k in keywords if k in description_lower]
        if matches:
            score = len(matches) / len(keywords) if keywords else 0.1
            scores[category] = score
            matched_keywords[category] = matches

    if not scores:
        return "Other", 0.3, []

    # Get category with highest score
    best_category = max(scores, key=scores.get)
    confidence = scores[best_category]

    return best_category, confidence, matched_keywords.get(best_category, [])

@router.post("/")
async def categorize(request: CategorizationRequest) -> CategorizationResponse:
    """
    Categorize an expense based on its description
    """
    try:
        if not request.description or len(request.description.strip()) == 0:
            raise HTTPException(status_code=400, detail="Description cannot be empty")

        category, confidence, keywords = categorize_expense(request.description)

        # Return lower confidence if below threshold
        if confidence < request.confidence_threshold:
            category = "Other"
            confidence = 0.3

        return CategorizationResponse(
            category=category,
            confidence=round(confidence, 2),
            keywords_matched=keywords
        )

    except Exception as e:
        logger.error(f"Categorization error: {str(e)}")
        raise HTTPException(status_code=500, detail="Categorization failed")

@router.get("/categories")
async def get_categories():
    """
    Get list of all available categories
    """
    return {
        "categories": [cat for cat in CATEGORY_KEYWORDS.keys() if cat != "Other"],
        "total": len(CATEGORY_KEYWORDS) - 1
    }

File: lib/ai/categorizer.ts
import { ExpenseCategory } from '../types';

const categoryKeywords: { [key in ExpenseCategory]: string[] } = {
  Food: [
    'coffee',
    'restaurant',
    'pizza',
    'burger',
    'lunch',
    'dinner',
    'breakfast',
    'food',
    'grocery',
    'supermarket',
    'cafe',
    'bakery',
    'delivery',
    'uber eats',
    'doordash',
    'grubhub',
    'milk',
    'bread',
    'fruit',
    'vegetable',
  ],
  Transport: [
    'bus',
    'taxi',
    'uber',
    'lyft',
    'gas',
    'petrol',
    'parking',
    'train',
    'subway',
    'metro',
    'fuel',
    'toll',
    'transit',
    'transportation',
    'flight',
    'airline',
  ],
  Education: [
    'school',
    'course',
    'tuition',
    'textbook',
    'book',
    'university',
    'college',
    'class',
    'lesson',
    'learning',
    'education',
    'udemy',
    'coursera',
    'study',
    'exam',
  ],
  Entertainment: [
    'movie',
    'cinema',
    'theater',
    'concert',
    'game',
    'netflix',
    'spotify',
    'hulu',
    'disney',
    'streaming',
    'ticket',
    'show',
    'entertainment',
    'podcast',
  ],
  Utilities: [
    'electricity',
    'water',
    'gas',
    'internet',
    'phone',
    'bill',
    'utility',
    'power',
    'broadband',
    'cable',
    'telecom',
    'provider',
  ],
  Shopping: [
    'mall',
    'shop',
    'store',
    'amazon',
    'ebay',
    'target',
    'walmart',
    'shopping',
    'clothes',
    'dress',
    'shoe',
    'fashion',
    'apparel',
    'retail',
    'purchase',
    'fashion',
  ],
  Health: [
    'doctor',
    'hospital',
    'pharmacy',
    'medicine',
    'health',
    'gym',
    'fitness',
    'yoga',
    'medical',
    'clinic',
    'dental',
    'dentist',
    'prescription',
    'healthcare',
  ],
  Other: ['misc', 'other', 'miscellaneous'],
};

export function categorizeExpense(description: string): ExpenseCategory {
  const lowerDescription = description.toLowerCase();

  // Score each category based on matching keywords
  const scores: { [key in ExpenseCategory]: number } = {
    Food: 0,
    Transport: 0,
    Education: 0,
    Entertainment: 0,
    Utilities: 0,
    Shopping: 0,
    Health: 0,
    Other: 0,
  };

  // Count keyword matches for each category
  Object.entries(categoryKeywords).forEach(([category, keywords]) => {
    keywords.forEach((keyword) => {
      if (lowerDescription.includes(keyword)) {
        scores[category as ExpenseCategory] += 1;
      }
    });
  });

  // Find category with highest score
  let maxScore = 0;
  let bestCategory: ExpenseCategory = 'Other';

  Object.entries(scores).forEach(([category, score]) => {
    if (score > maxScore) {
      maxScore = score;
      bestCategory = category as ExpenseCategory;
    }
  });

  return maxScore > 0 ? bestCategory : 'Other';
}

// Alternative: Use a more sophisticated pattern matching
export function categorizeExpenseAdvanced(description: string): {
  category: ExpenseCategory;
  confidence: number;
} {
  const category = categorizeExpense(description);

  const lowerDescription = description.toLowerCase();
  const keywords = categoryKeywords[category] || [];

  const matchCount = keywords.filter((k) => lowerDescription.includes(k)).length;
  const confidence = Math.min(100, (matchCount * 25) + 25); // Base 25% to 100%

  return { category, confidence };
}
